porch/pkg/engine/engine.go

```diff
@@ -42,18 +44,72 @@ type CaDEngine interface {
-	DeletePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, obj repository.PackageRevision) error
+	ListPackageRevisions(ctx context.Context, repositorySpec *configapi.Repository, filter repository.ListPackageRevisionFilter) ([]*PackageRevision, error)
+	CreatePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, obj *api.PackageRevision) (*PackageRevision, error)
+	UpdatePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, oldPackage *PackageRevision, old, new *api.PackageRevision) (*PackageRevision, error)
+	DeletePackageRevision(ctx context.Context, repositoryObj *configapi.Repository, obj *PackageRevision) error
 
-	ListPackages(ctx context.Context, repositorySpec *configapi.Repository, filter repository.ListPackageFilter) ([]repository.Package, error)
-	CreatePackage(ctx context.Context, repositoryObj *configapi.Repository, obj *api.Package) (repository.Package, error)
-	UpdatePackage(ctx context.Context, repositoryObj *configapi.Repository, oldPackage repository.Package, old, new *api.Package) (repository.Package, error)
-	DeletePackage(ctx context.Context, repositoryObj *configapi.Repository, obj repository.Package) error
+	ListPackages(ctx context.Context, repositorySpec *configapi.Repository, filter repository.ListPackageFilter) ([]*Package, error)
+	CreatePackage(ctx context.Context, repositoryObj *configapi.Repository, obj *api.Package) (*Package, error)
+	UpdatePackage(ctx context.Context, repositoryObj *configapi.Repository, oldPackage *Package, old, new *api.Package) (*Package, error)
+	DeletePackage(ctx context.Context, repositoryObj *configapi.Repository, obj *Package) error
+}
+
+type Package struct {
+	RepoPackage repository.Package
+}
+
+func (p *Package) GetPackage() *api.Package {
+	return p.RepoPackage.GetPackage()
+}
+
+func (p *Package) KubeObjectName() string {
+	return p.RepoPackage.KubeObjectName()
+}
+
+type PackageRevision struct {
+	RepoPackageRevision repository.PackageRevision
+	PackageRevisionMeta meta.PackageRevisionMeta
+}
+
+func (p *PackageRevision) GetPackageRevision() *api.PackageRevision {
+	repoPkgRev := p.RepoPackageRevision.GetPackageRevision()
+	var isLatest bool
+	if val, found := repoPkgRev.Labels[api.LatestPackageRevisionKey]; found && val == api.LatestPackageRevisionValue {
+		isLatest = true
+	}
+	repoPkgRev.Labels = p.PackageRevisionMeta.Labels
+	if isLatest {
+		if repoPkgRev.Labels == nil {
+			repoPkgRev.Labels = make(map[string]string)
+		}
+		repoPkgRev.Labels[api.LatestPackageRevisionKey] = api.LatestPackageRevisionValue
+	}
+	repoPkgRev.Annotations = p.PackageRevisionMeta.Annotations
+	return repoPkgRev
+}
+
+func (p *PackageRevision) KubeObjectName() string {
+	return p.RepoPackageRevision.KubeObjectName()
+}
+
+func (p *PackageRevision) GetResources(ctx context.Context) (*api.PackageRevisionResources, error) {
+	return p.RepoPackageRevision.GetResources(ctx)
+}
+
+type Function struct {
+	RepoFunction repository.Function
```

Comment:
Could we embed this `RepoFunction`, and likewise the `RepoPackage` in the above Package struct? It would reduce the amount of code needed in this file

Reply:
I want the new types to be the interface exposed outside of the Engine, and it is a pretty limited number of functions needed. Embedding the types would mean that the full interface of those types would be available. I considered adding interfaces for the types, but it seemed unnecessary at this point.